### ISSUE TITLE
Documentation examples

### ISSUE THREAD

**Issue by brucou** [gcanti/fp-ts#265]
Hi there.

First of all, congrats on this extensive library. I followed for some time your development effort and I must say, and saw the adjunction of documentation (mostly based on type signature), and I think that is great. 

However I think that users like me would benefit enormously from a documentation which also portrays examples. A good example in that aspect is Ramda's, which I guess you all know (http://ramdajs.com/docs). Leaving aside the bells and whistles of ramda's doc (playground, etc.), what is great is the examples and the categorization (Function, Map, List, Object). Current documentation for fp-ts consist mainly of the name of the function and the corresponding type, which makes the whole thing not very friendly to absorb for beginners.

A part of the docs already do some of this :  

- https://github.com/gcanti/io-ts
- https://github.com/gcanti/newtype-ts

Something I was looking at recently with a view to using it and I could not figure out exactly how to do so : 

- https://github.com/gcanti/fp-ts/blob/master/docs/api/md/Store.md
- https://github.com/gcanti/fp-ts/blob/master/docs/api/md/Traversable.md
  - That http://ramdajs.com/docs/#traverse reads better than `traverse: <F>(F: Applicative<F>) => <A, B>(f: (a: A) => HKT<F, B>, ta: HKT<T, A>) => HKT<F, HKT<T, B>>`

I can probably figure out all of what I need to figure out but my fear is that I would spend too much time doing so. I can see in my code that I can benefits from functional abstraction (for instance I have a contract library which check assertions and aggregate errors, all of which could benefit from the applicative programming tools you have in fp-ts). But with the perceived syntax, abstraction, and documentation barriers, I feel like it is safer to keep things in plain and simple js/ts. Even supposing I figure everything out in a relatively short time, my colleagues or whoever comes after me, will have to do the exact same job, which compounds the perceived productivity loss.

Now, I am aware that there are a HUGE number of functions, and functionalities addressed in this library, so it looks like a large quantity of work to add examples everywhere. I still wanted however to give you guys my feedback, as I think it is also interesting to hear about the people who are not using the library, and why.

When I am between projects and with more time, I will be happy to properly dedicate resources to see how to integrate this library in my work.

**Comment by raveclassic** [comment]
@brucou 

> Something I was looking at recently with a view to using it and I could not figure out exactly how to do so

The reason is that ramda and other fp libs use Hindley–Milner notation (which is common for Haskell) when fp-ts being designed for TypeScript uses TypeScript signatures to express the type.
I don't know which one would be the best, just drawing your attention.

**Comment by raveclassic** [comment]
related https://github.com/gcanti/fp-ts/issues/157

**Comment by gcanti** [comment]
> Maybe signatures could be duplicated with Haskell version in the docs?

@raveclassic Too much work and you likely end up out of sync. 

The distinctive feature of fp-ts is faking HKTs and abstract over them, no other lib is trying to do that. 
Writing a data structure like `Option` or `Either` is relatively trivial, 
defining `Traversable` and then `sequence` in a type-safe and usable manner is not.

However you end up with

- verbose signatures
- tons of overloadings

IME there are some informal rules to move back and forth though:

### Type constructors

```purescript
f a
```

-> `HKT`

```ts
HKT<F, A>
```

### Type classes

```purescript
class Functor f where
  map :: forall a b. (a -> b) -> f a -> f b
```

-> `interface`s

```ts
interface Functor<F> {
  map<A, B>(f: (a: A) => B, fa: HKT<F, A>): HKT<F, B>
}
```

### Instances

```purescript
instance semigroupString :: Semigroup String where
  append = concatString
```

-> static dictionaries

```ts
const semigroupString: Semigroup<string> = {
  concat: x => y => x + y
}
```

### Type contraints

```purescript
instance semigroupMaybe :: Semigroup a => Semigroup (Maybe a) where
  append Nothing y = y
  append x Nothing = x
  append (Just x) (Just y) = Just (x <> y)
```

-> functions with additional initial arguments and returning instances

```ts
const getSemigroup = <A>(S: Semigroup<A>): Semigroup<Option<A>> => {
  return {
    concat: x => y => x.fold(() => y, ax => y.fold(() => x, ay => some(S.concat(ax)(ay))))
  }
}
```

**Comment by raveclassic** [comment]
@gcanti Nice explanation actually, should it be included into readme?

**Comment by gcanti** [comment]
Sure, I'll gladly accept any PR which improves the documentation with such explanations and/or adds some examples.

I tried hard to follow those rules across the whole code base, for example here's the signature of `sequence` in PureScript

```purescript
sequence :: forall t a f. Applicative f => Traversable t => t (f a) -> f (t a)
```

and here's the fp-ts counterpart

```ts
//                Applicative f =>   Traversable t =>     t (f a) ->                 f (t a)
sequence<F, T>(F: Applicative<F>, T: Traversable<T>): <A>(tfa: HKT<T, HKT<F, A>>) => HKT<F, HKT<T, A>>
```

is almost a one-to-one translation ;)

**Comment by nghiepit** [comment]
`fp-ts` Not for everyone. Only for people with a lot of Typescript experience.
No examples, no playground. Just only the name of the function and the corresponding type.
When trying to use `fp-ts`, I cried a lot, even the most basic feature such as `Option`, `Either`, `pipe`.

**Comment by brucou** [comment]
@nghiepit  I can agree with that. A lot of the blame I think is on TypeScript in the first place. After trying to use a variety of TS functional programming libraries I came to the conclusion that there isn't much to be gained from using formal functional patterns like monad and so on in JS. For types I use JSDoc. There is far too much syntactic noise involved in TS types, and TS functional libraries only compound that. The resulting code is hard to read for everybody, No amount of type safety compensates for that.

But again, the team has done a great job I think over the years with the documentation.

**Comment by enricopolanski** [comment]
> `fp-ts` Not for everyone. Only for people with a lot of Typescript experience.
> No examples, no playground. Just only the name of the function and the corresponding type.
> When trying to use `fp-ts`, I cried a lot, even the most basic feature such as `Option`, `Either`, `pipe`.

You are more than welcome to contribute in developing a playground and examples, it would be great!
